test/jdk/java/lang/StringBuffer/ECoreIndexOf.java

```diff
@@ -0,0 +1,1370 @@
+        System.err.println("sourceBuffer.indexOf(emptyString, 99) fragment '" + emptyString + "' ("
+            + emptyString.length() + ") String = "
+            + sourceBuffer.toString() + " len Buffer = " + sourceBuffer.toString().length());
+        System.err.println("  naive = " + naiveFind(sourceBuffer.toString(), emptyString, 99) + ", IndexOf = "
+            + indexOfKernel(sourceBuffer, emptyString, 99));
+        failCount++;
+      }
+      if ((indexOfKernel(allAsBuffer.substring(1, 3), allAsBuffer.substring(5, 12)) != -1) ||
+          (-1 != naiveFind(allAsBuffer.substring(1, 3).toString(), allAsBuffer.substring(5, 12), 0))) {
+        System.err.println("allAsBuffer.substring(1, 3).indexOf(allAsBuffer.substring(5, 12)) fragment '"
+            + allAsBuffer.substring(5, 12) + "' ("
+            + allAsBuffer.substring(5, 12).length() + ") String = "
+            + allAsBuffer.substring(1, 3) + " len Buffer = " + allAsBuffer.substring(1, 3).length());
+        System.err.println(
+            "  naive = " + naiveFind(allAsBuffer.substring(1, 3).toString(), allAsBuffer.substring(5, 12), 0)
+                + ", IndexOf = " + indexOfKernel(allAsBuffer.substring(1, 3), allAsBuffer.substring(5, 12)));
+        failCount++;
+      }
+    }
+
+    report("Basic Test                   ", failCount);
+  }
+
+  // Note: it is possible although highly improbable that failCount will
+  // be > 0 even if everthing is working ok
+  private static void compareIndexOfLastIndexOf() {
+    int failCount = 0;
+    String sourceString;
+    StringBuffer sourceBuffer;
+    String targetString;
+
+    for (int i = 0; i < 10000; i++) {
+      do {
+        sourceString = generateTestString(99, 100);
+        sourceBuffer = new StringBuffer(sourceString);
+        targetString = generateTestString(10, 11);
+      } while (indexOfKernel(sourceString, targetString) != -1);
+
+      int index1 = generator.nextInt(100);
+      sourceBuffer = sourceBuffer.replace(index1, index1, targetString);
+
+      // extremely remote possibility of > 1 match
+      int matches = 0;
+      int index2 = -1;
+      while ((index2 = indexOfKernel(sourceBuffer, targetString, index2 + 1)) != -1)
+        matches++;
+      if (matches > 1)
+        continue;
+
+      if (indexOfKernel(sourceBuffer, targetString) != sourceBuffer.lastIndexOf(targetString))
+        failCount++;
+      sourceString = sourceBuffer.toString();
+      if (indexOfKernel(sourceString, targetString) != sourceString.lastIndexOf(targetString))
+        failCount++;
+    }
+
+    report("IndexOf vs LastIndexOf       ", failCount);
+  }
+
+  private static void compareStringStringBuffer() {
+    int failCount = 0;
+    boolean make_new = true;
+
+    String fragment = null;
+    StringBuffer testBuffer = null;
+    String testString = null;
+    int testIndex = 0;
+
+    failCount = indexOfKernel("", "");
+
+    for (int x = 0; x < 1000000; x++) {
```

Comment:
Should we be concerned about the increased run-time? Or does this execute 'quickly enough'

Reply:
Runs in well under 2 minutes.